### DISCUSSION TITLE
:gemoji: gets an unwanted slash when converting from AST to Markdown

### DISCUSSION THREAD

**Discussion by vrepsys** [micromark/micromark#151] [Q&A]
Hello all,

looking for some guidance with this issue I have.

The issue I'm facing is that remark-directive messes with how gemojis are rendered to from AST->Markdown.

Basically if there's a paragraph with a :gemoji:, 
1) It parses into an AST with a paragraph containing a :gemoji:.(Geat, just as documented!)
2) AST with the paragraph containing :gemoji: is rendered back to markdown, it now has a backslash in front of it /:gemoji: (Not good)

Is there a reason why it gets a backslash? I understand why a paragraph with :textDirective should be escaped, but :this: is not a text directive. Just want to understand.

```ts
const ast = fromMarkdown(":rocket:", {
  extensions: [directive()],
  mdastExtensions: [directiveFromMarkdown()],
});
console.log(JSON.stringify(ast.children[0].type)); // paragraph

const md = toMarkdown(ast.children[0], { extensions: [directiveToMarkdown()] });
console.log(md); // \:rocket:
```

I dug a bit into `mdast-util-directive` code, trying to find answers.
It seems like this part might have something to do with how AST is converted to Markdown

```js
export function directiveToMarkdown() {
  return {
    unsafe: [
      {
        character: '\r',
        inConstruct: ['leafDirectiveLabel', 'containerDirectiveLabel']
      },
      {
        character: '\n',
        inConstruct: ['leafDirectiveLabel', 'containerDirectiveLabel']
      },
      {
        before: '[^:]',
        character: ':',
        after: '[A-Za-z]',
        inConstruct: ['phrasing']
      },
      {atBreak: true, character: ':', after: ':'}
    ],
    handlers: {
      containerDirective: handleDirective,
      leafDirective: handleDirective,
      textDirective: handleDirective
    }
  }
}
```

Couldn't figure out which part exactly cause the slash to appear. If anyone has recommendations for how to start learning how micromark extensions work please share! I've already read the micromark README.

Sorry for long text, too late in the evening to write short and concise:)

**Comment by ChristianMurphy** [comment]
Welcome @vrepsys! 👋 

Linking a related discussion, it feels like this may be similar to or the same as https://github.com/orgs/remarkjs/discussions/651

Where because of the overlap in syntax between gemoji and text directive, an escape is added by remark directive to ensure serialization does not accidentally add a directive where there should not be one.

**Reply by vrepsys** [reply]
Thanks @ChristianMurphy, useful context! 

My understanding is that :this: doesn't need an escape baskslash, becaue it's not a directive. But that's probablly a side effect of escaping :real directives.

**Reply by wooorm** [reply]
Correct, but that requires a lot of context. When looking at `:`, we’d need to know many future characters, to find that other `:`, with a valid word in between. We don’t have that.
We only have the previous and next character surrounding `:`

**Comment by wooorm** [comment]
> AST with the paragraph containing :gemoji: is rendered back to markdown, it now has a backslash in front of it /:gemoji: (Not good)

GH supports this. `:sweat_smile:` -> :sweat_smile:, `:sweat\_smile:` -> :sweat\_smile:, `\:+1:` -> \:+1:

---

Why doesn’t your setup support this?

**Reply by vrepsys** [reply]
I'm working on a markdown editor that takes people's markdown from their Github repos, lets them edit it, and then push back to Github (dhub.dev).

So even if user didn't make any changes to their md file, all gemojis get unnecessarily escaped, which looks really bad in Github compare and also reduces markdown's readability (sometimes there are lots and lots of gemojis.. like here for example https://raw.githubusercontent.com/facebook/docusaurus/main/CHANGELOG.md)

But I agree, parser-wise everything still works with the backslash.

**Reply by wooorm** [reply]
in that case, why use directives? GH doesn’t support ’em?

**Reply by wooorm** [reply]
like, because that’s important. Directives interfere with gemoji. I patched stuff to make it mostly work on the directive side. But when GH actually supports directives, they’ll have some thinking to do on how to actually implement the different. And then we have a luxury problem: directives are popular and I’ll gladly update everything to match it well!

**Reply by vrepsys** [reply]
I need to parse containerDirectives because this is how [admonitions/callouts](https://docusaurus.io/docs/markdown-features/admonitions) work in Docusaurus. 

Don't really need text and leaf directives at this moment, but I haven't found an elegant way to switch them off.

I'm basically trying to implement mdx->ast->mdx for Docusaurus markdown.

**Reply by wooorm** [reply]
There’s no way to know if `:a` will turn into a directive or an emoji. If docusaurus support directives, which I think they do, they will be able to turn stuff into directives too

**Reply by vrepsys** [reply]
Yes, they use the remark-directive, haven't seen any uses of text directives though (doesn't mean there are none...)

Not sure I understood your point about not knowing what :a will turn into.
From where I stand if \:g: or :g is inside a text node in an AST -- definitely neither a directive nor a gemoji.

**Reply by wooorm** [reply]
Assuming you are a JS formatter/serializer, and you have a string literal of `a "b" c`, you can’t format it as `"a "b" c"` right? That breaks.

Same for us, if someone generates `a *b* c` in a text node, that’s going to turn into emphasis. So it doesn’t work.

Now, imagine a language where anyone has arbitrary extensions (markdown). It gets complex. Particularly because this language doesn’t look at a few characters, but is based on a ton of context.

**Reply by vrepsys** [reply]
I think I'm starting to get it. You're saying that if we don't add a backslash when stringifying a text node with :this: -- there might be issues when trying to parse it correctly next time.

But right now, there don't seem to be issues with distinguishing :this from :this:, seems to work correctly.

**Reply by wooorm** [reply]
a) Yup!
b) Interesting. Well, it depends I guess on whether `[` and `{` are escaped when directives are enabled? When there’s a `[^:]:\w`, it currently gets escaped by the directives. If a text node includes `:a[]` or `:a{}`, and it serializes like that unescaped, we have a problem. It looks like the requirement for `[]` or `{}` for texts is also recent, and isn’t in https://github.com/syntax-tree/mdast-util-directive/blob/5dd6e97d7fa3089f25a9af7b4ee382cc9dacdd9b/lib/index.js#L289 yet?

**Reply by vrepsys** [reply]
Yep, `:this:` parses as a paragraph `:this[x]:` and `:this[x]{a='b'}:` parse into directive nodes with a text node containing the `:` following.

**Reply by wooorm** [reply]
My q is about how they serialize as well! Whtehr the brackets and braces get encoded or not!

**Reply by vrepsys** [reply]
Good morning!

> it looks like the requirement for [] or {} for texts is also recent

What requirements do you have in mind? Github? Do you have a link?

I'm now wondering if parsing of `\:gemoji:` as a gemoji node is correct behaviour. Shouldn't the slash escape the first semicolon and cause the whole thing to render as text?

https://spec.commonmark.org/0.30/#backslash-escapes

**Reply by wooorm** [reply]
> What requirements do you have in mind? Github? Do you have a link?

`micromark-extension-directive` changed. It used to see `:a` as a text directive. It no longer does.
`mdast-util-directive` is perhaps unaware of this change.

> I'm now wondering if parsing of `\:gemoji:` as a gemoji node is correct behaviour. Shouldn't the slash escape the first semicolon and cause the whole thing to render as text?

You can take that up with GitHub (theoretically, because practically they don’t really work on their markdown things).
Gemoji are not an extension to markdown.
They are instead something that is post processed, transformed on the HTML tree.
At that point, escapes no longer exist.

**Reply by vrepsys** [reply]
> `micromark-extension-directive` changed. It used to see `:a` as a text directive. It no longer does.
> `mdast-util-directive` is perhaps unaware of this change.

Interesting! Do I understand correctly that if `:a` is no longer a directive `:a:` doesn't have to be escaped any longer? It's just that it hasn't been implemented in `mdast-util-directive` yet.

> They are instead something that is post processed, transformed on the HTML tree.

Got it

**Reply by wooorm** [reply]
Perhaps. But we don’t know if `:a` is followed by `[` or a `{`. If we’re not escaping `:`, we need to escape all `[` and `{`s. That might also have downsides.
